### PR TITLE
Reference Tests and Breaking Change: Optional nullable fields are now null instead of undefined

### PR DESCRIPTION
**Pull request by wilwade**
Problem
=======

We wanted to add tests for all the tests in https://github.com/apache/parquet-testing

### Discovered Bugs
- We treated nulls as undefined, but others don't
- We incorrectly processed dictionary_page_offset >= 0 instead of only > 0

Solution
========

- Added new test that automatically tests all files: `test/reference-test/read-all.test.ts`
- Fixed found bugs

with @shannonwells 

Steps to Verify:
----------------
1. Run the tests
1. Comment out the bug fixes and see reference test files fail

### CODE REVIEW

File: test/reference-test/read-all.test.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  'non_hadoop_lz4_compressed.parquet', // ZSTD unsupported
+  'rle_boolean_encoding.parquet', // BUG?: https://github.com/LibertyDSNP/parquetjs/issues/113
+  'datapage_v2.snappy.parquet', // DELTA_BINARY_PACKED unsupported
+  'delta_binary_packed.parquet', // DELTA_BINARY_PACKED unsupported
+  'delta_byte_array.parquet', // DELTA_BYTE_ARRAY unsupported
+  'delta_encoding_optional_column.parquet', // DELTA_BINARY_PACKED unsupported
+  'delta_encoding_required_column.parquet', // DELTA_BINARY_PACKED unsupported
+  'delta_length_byte_array.parquet', // ZSTD unsupported, DELTA_BINARY_PACKED unsupported
+  'float16_nonzeros_and_nans.parquet', // missing option: typeLength (required for FIXED_LEN_BYTE_ARRAY)
+  'float16_zeros_and_nans.parquet', // missing option: typeLength (required for FIXED_LEN_BYTE_ARRAY)
```
Comment (wilwade) [inline test/reference-test/read-all.test.ts:28]: These can be removed once #108 is rebased on it and will work as a test to prove it is working.

File: lib/shred.ts [hunk 2]
```diff
@@ -227,6 +227,8 @@ function materializeRecordField(record: Record<string, unknown>, branch: Array<P
   const node = branch[0];
 
   if (dLevel < node.dLevelMax) {
+    // This ensures that nulls are correctly processed
+    record[node.name] = value;
```
Comment (wilwade) [inline lib/shred.ts:231]: This is the change that makes null values shred to null instead of undefined.

File: lib/reader.ts [hunk 3]
```diff
@@ -672,8 +672,9 @@ export class ParquetEnvelopeReader {
       num_values: metadata.num_values
     });
 
-    if (metadata.dictionary_page_offset) {
-      const offset = +metadata.dictionary_page_offset;
+    // If this exists and is greater than zero then we need to have an offset
+    if (metadata.dictionary_page_offset && +metadata.dictionary_page_offset > 0) {
+      const offset: number = +metadata.dictionary_page_offset;
```
Comment (wilwade) [inline lib/reader.ts:677]: Based in part around the information in https://issues.apache.org/jira/browse/PARQUET-1402

File: test/shred.js
Comment (wilwade) [inline test/shred.js:1]: Fixes for using null instead of undefined

File: test/test-files.js
Comment (wilwade) [inline test/test-files.js:1]: Fixes for using null instead of undefined

File: test/integration.js
Comment (wilwade) [inline test/integration.js:1]: Fixes for using null instead of undefined

**Review by shannonwells — APPROVED** [review]
Works on my machine

**Pull request merged**